### PR TITLE
Non-blocking primary relocation hand-off

### PR DESCRIPTION
**Pull request by ywelsch**
Primary relocation and indexing concurrently can currently lead to a deadlock situation as indexing operations are blocked on a (bounded) thread pool during the hand-off phase between old and new primary. This PR replaces blocking of indexing operations by putting operations that cannot be executed during relocation hand-off in a queue to be executed once relocation completes.

Relates to #18553, #15900.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -235,17 +237,48 @@ public void messageReceived(final Request request, final TransportChannel channe
...
         @Override
-        public void messageReceived(Request request, TransportChannel channel, Task task) throws Exception {
+        public void messageReceived(Request request, TransportChannel channel, Task task) {
             ReplicationTask replicationTask = (ReplicationTask) task;
-            boolean success = false;
-            PrimaryShardReference primaryShardReference = getPrimaryShardReference(request.shardId());
-            try {
-                if (primaryShardReference.isRelocated()) {
+
+            Callback<Throwable> onFailure = t -> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: all these small call backs make me think we need asyncPrimaryAction which has the channel and the task as fields and implements ActionListener<PrimaryShardReference> so we can pass it as a value to the call backs?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -264,29 +297,12 @@ public void handleException(TransportException exp) {
...
-            } finally {
-                if (success == false) {
-                    primaryShardReference.close();
+                } catch (Throwable t) {
+                    onFailure.handle(t);
                 }
-            }
+            };
+
+            acquirePrimaryShardReference(request.shardId(), onReferenceAcquired, onRelocated, onFailure);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: why can't we use an ActionListener<PrimaryShardReference> call back here? I think we can deal with the relocation check in the same way did before, within the onResponse method?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I would very much prefer an `AsyncPrimaryAction` (especially so that it implements AbstractRunnable, making failures simpler).

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -436,20 +437,27 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
...
 
-    public IndexShard relocated(String reason) throws IndexShardNotStartedException {
-        try (Releasable block = suspendableRefContainer.blockAcquisition()) {
-            // no shard operation locks are being held here, move state from started to relocated
-            synchronized (mutex) {
-                if (state != IndexShardState.STARTED) {
-                    throw new IndexShardNotStartedException(shardId, state);
+    public void relocated(String reason) throws IllegalIndexShardStateException, InterruptedException {
+        try {
+            indexShardOperationsLock.blockOperations(30, TimeUnit.MINUTES, () -> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: ++, one less releasable to remember to free.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            } else {
+                // relocation hand-off in progress, store onRelocated callback to trigger after hand-off finished
+                synchronized (mutex) {
+                    if (blockOperationsInProgress) {
+                        // we know this will be released by call to releaseDelayedOperations
+                        if (delayedOperations == null) {
+                            delayedOperations = new ArrayList<>();
+                        }
+                        if (executor != null) {
+                            delayedOperations.add(new ThreadedActionListener(logger, threadPool, executor, onAcquired, true));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: forcing execution should be a parameter for now imo - I know we want/maybe/potentially change how we deal with replicas and queues, but for now I rather not change semantics and have primary ops non-forced and replicas ops forced.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: although I like it, it means the implementation on the indexShardOperationsLock can not use simple synchornized methods  because their duration is unclear . if this returned a releasble like it used to, I *think* all ops over there can be synchorinze and we won't need fancy non-blocking implementations with `while(true)` . If so, I think that's the right trade off to roll this back?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 5]
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
...
-
         cancellableThreads.execute(recoveryTarget::finalizeRecovery);
 
         if (isPrimaryRelocation()) {
+            cancellableThreads.checkForCancel();
+            logger.trace("[{}][{}] performing relocation hand-off to {}", indexName, shardId, request.targetNode());
+            try {
+                cancellableThreads.execute(() -> shard.relocated("to " + request.targetNode()));
+            } catch (Throwable e) {
+                logger.debug("[{}][{}] completing relocation hand-off to {} failed", indexName, shardId, request.targetNode(), e);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I think the exception is not in the right place?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java [hunk 6]
```diff
@@ -263,12 +263,6 @@ public RecoveryResponse newInstance() {
                 return;
             }
 
-            if (cause instanceof IndexShardClosedException) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: why did you remove this?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 7]
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
...
+                throw e;
+            }
             /**
              * if the recovery process fails after setting the shard state to RELOCATED, both relocation source and
              * target are failed (see {@link IndexShard#updateRoutingEntry}).
              */
-            try {
-                shard.relocated("to " + request.targetNode());
-            } catch (IllegalIndexShardStateException e) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: ++ on removing this catch. Not true any more

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java [hunk 8]
```diff
@@ -136,7 +136,7 @@ public ActionRequestValidationException validate() {
         @Override
         protected void doExecute(Task task, final Request request, ActionListener<Response> listener) {
             // remove unneeded threading by wrapping listener with SAME to prevent super.doExecute from wrapping it with LISTENER
-            super.doExecute(task, request, new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.SAME, listener));
+            super.doExecute(task, request, new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.SAME, listener, false));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: I wonder if we should make false be the default with an overload?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java (hunk 6 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: It's dead code. A few lines above we match the super class
`if (cause instanceof IllegalIndexShardStateException...`

**Comment by bleskes** [comment]
Thx @ywelsch . I left some comments that I think will simplify things. My main concern here is the extra IndexShardOperationsLock wrapper around `SuspendableRefContainer`. I'm not sure we need an extra abstraction instead of making SuspendableRefContainer implement the API we need (or just rename it). It makes things more complex, for example, IndexShardOperationsLock assumes that the only reason why tryAcquire can fail is that a block operation is going on. This is true now, but only because we use Integer.MAX_VALUE as the total amount of operations. Some one can change that and not realize the implications it has.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Good catch. I was surprised to see that this was not caught by the logger usage checker. After digging I noticed that the checker was accidentally disabled since it was added. I've opened #19126 to address this.

**Comment by ywelsch** [comment]
@bleskes I've updated the PR with the following main changes:
- introduced `AsyncPrimaryAction` and use the original flow of the method to distinguish `isRelocated` on the `PrimaryShardReference`.
- inlined `SuspendableRefContainer` into `IndexShardOperationsLock`.
- simplified lock acquisition retry (eliminating the loop) to remove some of the non-blocking fanciness.
Please have another look.

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java (hunk 8 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: It's only used in 3 places, so not worth the change imho.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 9]
```diff
@@ -157,7 +160,7 @@ protected void resolveRequest(MetaData metaData, IndexMetaData indexMetaData, Re
 
     /**
      * Synchronous replica operation on nodes with replica copies. This is done under the lock form
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: form -> from

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 10]
```diff
@@ -746,17 +789,29 @@ protected PrimaryShardReference getPrimaryShardReference(ShardId shardId) {
             throw new ReplicationOperation.RetryOnPrimaryException(indexShard.shardId(),
                 "actual shard is not a primary " + indexShard.routingEntry());
         }
-        return new PrimaryShardReference(indexShard, indexShard.acquirePrimaryOperationLock());
+
+        ActionListener<Releasable> onAcquired = new ActionListener<Releasable>() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: it's a shame we have to wrap the call back here, but I don't see how to simplify this without doing something not nice instead, like making the PrimaryShardRefernce having a non-final releasable field...

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            onAcquired.onResponse(releasable);
+        } else {
+            synchronized (mutex) {
+                if (blockOperationsInProgress) {
+                    // blockOperations is executing, this operation will be retried by blockOperations once it finishes
+                    if (delayedOperations == null) {
+                        delayedOperations = new ArrayList<>();
+                    }
+                    if (executor != null) {
+                        delayedOperations.add(new ThreadedActionListener(logger, threadPool, executor, onAcquired, forceExecution));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: should be ThreadedActionListener<>

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+     * operations that are occurring in the meanwhile and runs them once onBlocked has executed.
+     *
+     * @param timeout the maximum time to wait for the in-flight operations block
+     * @param timeUnit the time unit of the {@code timeout} argument
+     * @param onBlocked the action to run once the block has been acquired
+     * @throws IllegalStateException if operation is called while running
+     * @throws InterruptedException if calling thread is interrupted
+     * @throws TimeoutException if timed out waiting for in-flight operations to finish
+     */
+    public void blockOperations(long timeout, TimeUnit timeUnit, Runnable onBlocked) throws InterruptedException, TimeoutException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: I this this can simplified even further : https://gist.github.com/bleskes/0bf520c969eaa9542b1deefb4a8e1d5a  (also note the test fixes, which are unrelated)

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 13]
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
         cancellableThreads.checkForCancel();
         StopWatch stopWatch = new StopWatch().start();
         logger.trace("[{}][{}] finalizing recovery to {}", indexName, shardId, request.targetNode());
-
-
         cancellableThreads.execute(recoveryTarget::finalizeRecovery);
 
         if (isPrimaryRelocation()) {
+            cancellableThreads.checkForCancel();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: executes already checks for cancel. No need.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: I wonder if we should move it up above the IllegalIndexShardStateException, as we don't want to retry in this case (although it's not that bad).

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java (hunk 8 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: k

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 14]
```diff
@@ -629,26 +629,15 @@ public void execute() throws Exception {
                     }
                 };
             }
-        };
-        try {
-            primaryPhase.messageReceived(request, createTransportChannel(listener), task);
-        } catch (ElasticsearchException e) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: awesome

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+    @AfterClass
+    public static void shutdownThreadPool() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    @Before
+    public void createIndexShardOperationsLock() {
+         block = new IndexShardOperationsLock(logger, threadPool);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: I wonder if we want to add an `@After` rule that checks that all semaphore permits are back.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        blockingThread.start();
+
+        for (Thread thread : operationThreads) {
+            thread.start();
+        }
+
+        for (PlainActionFuture<Releasable> future : futures) {
+            Releasable releasable = future.get(1, TimeUnit.MINUTES);
+            assertNotNull(releasable);
+            releasable.close();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: this causes a deadlock when the operations that are blocked by the block are first in the future list while the last part of the future list prevents the block from being acquired. See the gist I made.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+
+        expectThrows(IllegalStateException.class, () -> block.blockOperations(1, TimeUnit.MINUTES, () -> {}));
+
+        releaseBlock.countDown();
+        blockingThread.join();
+    }
+
+    public void testOperationsInvokedImmediatelyIfNoBlock() throws ExecutionException, InterruptedException {
+        PlainActionFuture<Releasable> future = new PlainActionFuture<>();
+        block.acquire(future, ThreadPool.Names.GENERIC, true);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: maybe pass an executor that doesn't exist and will blow up?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 18]
```diff
@@ -1534,29 +1542,36 @@ private final EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
             IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING.get(indexSettings.getSettings()), refreshListeners);
     }
 
-    public Releasable acquirePrimaryOperationLock() {
+    /**
+     * Acquire a primary operation lock whenever the shard is ready for indexing. If the lock is directly available, the provided
+     * ActionListener will be called on the calling thread. During relocation hand-off, lock acquisition can be delayed. The provided
+     * ActionListener will then be called using the provided executor.
+     */
+    public void acquirePrimaryOperationLock(ActionListener<Releasable> onLockAcquired, String executor) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: maybe call the executor , executorOnDelay?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+     * Acquires a lock whenever lock acquisition is not blocked. If the lock is directly available, the provided
+     * ActionListener will be called on the calling thread. During calls of {@link #blockOperations(long, TimeUnit, Runnable)}, lock
+     * acquisition can be delayed. The provided ActionListener will then be called using the provided executor once blockOperations
+     * terminates.
+     *
+     * @param onAcquired ActionListener that is invoked once acquisition is successful or failed
+     * @param executor executor to use for delayed call
+     * @param forceExecution whether the runnable should force its execution in case it gets rejected
+     */
+    public void acquire(ActionListener<Releasable> onAcquired, String executor, boolean forceExecution) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: nit: use executorOnDelay as a name?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+        releaseBlock.countDown();
+
+        future.get(1, TimeUnit.MINUTES).close();
+
+        blockingThread.join();
+    }
+
+    protected Thread blockAndWait(final CountDownLatch releaseBlock) throws InterruptedException {
+        CountDownLatch blockAcquired = new CountDownLatch(1);
+        Thread blockingThread = new Thread() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: just use the thread pool?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+        blockAcquired.await();
+        return blockingThread;
+    }
+
+    public void testOperationsDelayedIfBlockFailure() throws ExecutionException, InterruptedException, TimeoutException {
+        CountDownLatch blockAcquired = new CountDownLatch(1);
+        CountDownLatch releaseBlock = new CountDownLatch(1);
+
+        final IndexShardClosedException exception = new IndexShardClosedException(new ShardId("blubb", "id", 0));
+        Thread blockingThread = new Thread() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: use thread pool?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+        future.get(1, TimeUnit.MINUTES).close();
+
+        blockingThread.join();
+    }
+
+    public void testActiveOperationsCount() throws ExecutionException, InterruptedException {
+        PlainActionFuture<Releasable> future1 = new PlainActionFuture<>();
+        block.acquire(future1, ThreadPool.Names.GENERIC, true);
+        assertTrue(future1.isDone());
+        assertThat(block.getActiveOperationsCount(), equalTo(1));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: I'm confused by this - how is the active operation 1?

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: Oh. You need to explicitly close the op.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 23]
```diff
@@ -898,12 +918,47 @@ public void testLockingBeforeAndAfterRelocated() throws Exception {
...
+            try {
+                latch.await();
+                shard.relocated("simulated recovery");
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        recoveryThread.start();
+        latch.countDown();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: it seems that the latch is useless? or maybe you wanted to enable the recovery half way through the "indexing operations"?  Also, this suffers from the same racing condition - if the ops that are now in flight when the relocation comes in are not first in the onLockAcquiredActions list, we will have a deadlock.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java (hunk 16 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: right

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 24]
```diff
@@ -436,20 +437,27 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
...
-        try (Releasable block = suspendableRefContainer.blockAcquisition()) {
-            // no shard operation locks are being held here, move state from started to relocated
-            synchronized (mutex) {
-                if (state != IndexShardState.STARTED) {
-                    throw new IndexShardNotStartedException(shardId, state);
+    public void relocated(String reason) throws IllegalIndexShardStateException, InterruptedException {
+        try {
+            indexShardOperationsLock.blockOperations(30, TimeUnit.MINUTES, () -> {
+                // no shard operation locks are being held here, move state from started to relocated
+                synchronized (mutex) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can we assert that there are no ops in flight?

File: core/src/test/java/org/elasticsearch/recovery/RelocationIT.java [hunk 25]
```diff
@@ -433,22 +433,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
         logger.info("red nodes: {}", redFuture.get());
-        ensureStableCluster(4);
+        ensureStableCluster(halfNodes * 2);
 
         assertAcked(prepareCreate("test").setSettings(Settings.builder()
             .put("index.routing.allocation.exclude.color", "blue")
-            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
-            .put(indexSettings())));
+            .put(indexSettings())
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0) // TODO: set this to randomInt(halfNodes - 1) once replica data loss is fixed
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RelocationIT.java]: can we open an issue for this and link to it with norelease so we won't forget?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 26]
```diff
@@ -1080,7 +1080,11 @@ private void assertShardIndexCounter() {
             IndicesService indexServices = getInstance(IndicesService.class, nodeAndClient.name);
             for (IndexService indexService : indexServices) {
                 for (IndexShard indexShard : indexService) {
-                    assertThat("index shard counter on shard " + indexShard.shardId() + " on node " + nodeAndClient.name + " not 0", indexShard.getActiveOperationsCount(), equalTo(0));
+                    try {
+                        assertBusy(() -> assertThat("index shard counter on shard " + indexShard.shardId() + " on node " + nodeAndClient.name + " not 0", indexShard.getActiveOperationsCount(), equalTo(0)));
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: why is this needed?

**Comment by bleskes** [comment]
@ywelsch I liked it! left a bunch of comments.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java (hunk 12 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: ok, I was afraid of lock contention by adding synchronization to the acquire method. We can try it your way and see how it impacts benchmarks.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: If you prefer, in the indexing method we can do tryAcquire first out of lock and go under lock and try again, if failed.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java (hunk 26 above)
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: The assertion trips on my CI when I don't put assertBusy() there. I'll have to investigate why.

Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: ok, I figured this one out. We have to release the shard operation lock before we respond to the caller, otherwise we might still hold on to the lock (for a very short time) after tests are finished.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java (hunk 12 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: Let's try it the simple way first (tryAcquire under lock) and see if it has an impact on indexing performance. Just FYI, I've adapted your implementation of `acquire` as it was holding onto the mutex when calling `onAcquired.onResponse(releasable)`.

**Comment by ywelsch** [comment]
@bleskes I've pushed a new set of changes addressing your comments. I've also added a `close` method to IndexShardOperationLock. This is not strictly necessary but makes operations fail faster when the shard is closed.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 27]
```diff
@@ -235,12 +238,33 @@ public void messageReceived(final Request request, final TransportChannel channe
...
+        @Override
+        protected void doRun() throws Exception {
+            acquirePrimaryShardReference(request.shardId(), this);
+        }
+
+        @Override
+        public void onResponse(PrimaryShardReference primaryShardReference) {
             try {
                 if (primaryShardReference.isRelocated()) {
+                    primaryShardReference.close(); // release shard operation lock as soon as possible
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: do we have a test that test that double closing on the shard op lock is OK?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: ping?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java [hunk 28]
```diff
@@ -0,0 +1,196 @@
...
+            };
+            futures.add(future);
+            operationThreads.add(thread);
+        }
+
+        for (Thread thread : operationThreads) {
+            thread.start();
+        }
+
+        threadPool.generic().execute(() -> {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: I think this can move to before the threads start -  will be nastier

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java (hunk 16 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: I think this is still a problem, no?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 27 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: yes, here: https://github.com/elastic/elasticsearch/pull/19013/files/a8a187b6e7f04c1621b1fa1db4f29f8832b91fc6#diff-322f042a012dbd1c09a9b43f40636655R180

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java [hunk 29]
```diff
@@ -0,0 +1,166 @@
...
+     * ActionListener will be called on the calling thread. During calls of {@link #blockOperations(long, TimeUnit, Runnable)}, lock
+     * acquisition can be delayed. The provided ActionListener will then be called using the provided executor once blockOperations
+     * terminates.
+     *
+     * @param onAcquired ActionListener that is invoked once acquisition is successful or failed
+     * @param executorOnDelay executor to use for delayed call
+     * @param forceExecution whether the runnable should force its execution in case it gets rejected
+     */
+    public void acquire(ActionListener<Releasable> onAcquired, String executorOnDelay, boolean forceExecution) {
+        if (closed) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: can we add a little test for the is new behavior?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 27 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: ++

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java (hunk 6 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: I'm not sure we should change this as part of this PR (I recall some issues when I tried that, will have to check again). I can leave this dead code for now and add a TODO instead.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java (hunk 16 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: uh, this change must haven been lost, will fix it.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java]: I'm fine with removing it - it's useless . I'm ok with just checking it further after the PR is done.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java (hunk 29 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java]: done in ae11e7e

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java (hunk 28 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java]: ok, done in ae11e7e

**Comment by bleskes** [comment]
LGTM. Thx @ywelsch
